File: simple_server.py
import http.server
import socketserver
import json
import os
import requests
from datetime import datetime
import urllib.parse
import cgi
import http.client
import base64

PORT = int(os.environ.get('PORT', 3000))

# Default colors to use when OpenAI API is not available
DEFAULT_COLORS = ["#FFD166", "#06D6A0", "#118AB2", "#EF476F", "#073B4C"]

class SimpleHTTPRequestHandler(http.server.BaseHTTPRequestHandler):
    def _set_headers(self, status_code=200, content_type='application/json'):
        self.send_response(status_code)
        self.send_header('Content-Type', content_type)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()

    def do_OPTIONS(self):
        self._set_headers(204)
        
    def do_POST(self):
        content_length = int(self.headers['Content-Length'])
        post_data = self.rfile.read(content_length)
        try:
            request_data = json.loads(post_data)
        except json.JSONDecodeError:
            self._set_headers(400)
            response = {'error': 'Invalid JSON'}
            self.wfile.write(json.dumps(response).encode())
            return
            
        parsed_path = urllib.parse.urlparse(self.path)
        
        if parsed_path.path == '/api/palette/generate':
            self._handle_generate_palette(request_data)
        elif parsed_path.path == '/api/palette/analyze':
            self._handle_analyze_image(request_data)
        else:
            self._set_headers(404)
            response = {
                'error': 'Not Found',
                'message': f'The POST endpoint {self.path} is not supported',
                'timestamp': datetime.now().isoformat()
            }
            self.wfile.write(json.dumps(response).encode())
    
    def _handle_generate_palette(self, data):
        # Extract parameters
        mood = data.get('mood', 'happy')
        description = data.get('description', '')
        count = data.get('count', 5)
        
        # Get OpenAI API key
        api_key = os.environ.get('OPENAI_API_KEY')
        
        if not api_key:
            self._set_headers(200)
            response = {
                'colors': DEFAULT_COLORS,
                'explanation': 'A default color palette (OpenAI API key not provided)',
                'generated': False
            }
            self.wfile.write(json.dumps(response).encode())
            return
            
        try:
            # Format the request for OpenAI
            moodDesc = f"{mood} mood and this description: '{description}'" if description else f"{mood} mood"
            prompt = f"Generate a cohesive color palette of {count} colors that represents a {moodDesc}. The colors should work well together and convey the right emotional tone."
            
            system_prompt = f"""You are a professional color theory expert and designer who creates perfect color palettes based on moods and emotions. 
            Return a JSON object with these keys: 
            "colors" (array of exactly {count} hex color codes like "#RRGGBB"), and 
            "explanation" (a brief description of the palette and how it relates to the requested mood).
            Ensure all colors work well together, have good contrast ratios when appropriate, and truly capture the essence of the requested mood."""
            
            # Call OpenAI API
            url = "https://api.openai.com/v1/chat/completions"
            headers = {
                "Content-Type": "application/json",
                "Authorization": f"Bearer {api_key}"
            }
            
            payload = {
                "model": "gpt-4o",
                "messages": [
                    {"role": "system", "content": system_prompt},
                    {"role": "user", "content": prompt}
                ],
                "temperature": 0.7,
                "max_tokens": 1000,
                "response_format": {"type": "json_object"}
            }
            
            response = requests.post(url, headers=headers, json=payload)
            response_data = response.json()
            
            # Process the OpenAI response
            if 'choices' in response_data and len(response_data['choices']) > 0:
                content = response_data['choices'][0]['message']['content']
                palette = json.loads(content)
                
                # Ensure we have the right number of colors
                if 'colors' in palette and len(palette['colors']) != count:
                    palette['colors'] = palette['colors'][:count]
                    
                    # If we still don't have enough, add some default colors
                    while len(palette['colors']) < count:
                        palette['colors'].append(DEFAULT_COLORS[len(palette['colors']) % len(DEFAULT_COLORS)])
                
                # Add generated flag
                palette['generated'] = True
                
                self._set_headers(200)
                self.wfile.write(json.dumps(palette).encode())
            else:
                # Return default if there's an issue with the OpenAI response
                self._set_headers(200)
                response = {
                    'colors': DEFAULT_COLORS,
                    'explanation': f"A default palette for '{mood}' mood (OpenAI API provided an invalid response)",
                    'generated': False
                }
                self.wfile.write(json.dumps(response).encode())
                
        except Exception as e:
            print(f"Error generating palette: {str(e)}")
            self._set_headers(500)
            response = {
                'error': f"Failed to generate color palette: {str(e)}",
                'colors': DEFAULT_COLORS,
                'explanation': f"A default palette (error occurred during generation)",
                'generated': False
            }
            self.wfile.write(json.dumps(response).encode())
    
    def _handle_analyze_image(self, data):
        # Extract parameters
        image_base64 = data.get('image', '')
        prompt = data.get('prompt', 'What colors are prominent in this image?')
        
        # Get OpenAI API key
        api_key = os.environ.get('OPENAI_API_KEY')
        
        if not api_key or not image_base64:
            self._set_headers(400)
            response = {
                'error': 'Missing API key or image data',
                'analysis': 'Unable to analyze image'
            }
            self.wfile.write(json.dumps(response).encode())
            return
            
        try:
            # Call OpenAI API with vision model
            url = "https://api.openai.com/v1/chat/completions"
            headers = {
                "Content-Type": "application/json",
                "Authorization": f"Bearer {api_key}"
            }
            
            payload = {
                "model": "gpt-4o",
                "messages": [
                    {
                        "role": "user",
                        "content": [
                            {
                                "type": "text",
                                "text": prompt
                            },
                            {
                                "type": "image_url",
                                "image_url": {
                                    "url": f"data:image/jpeg;base64,{image_base64}"
                                }
                            }
                        ]
                    }
                ],
                "max_tokens": 500
            }
            
            response = requests.post(url, headers=headers, json=payload)
            response_data = response.json()
            
            # Process the OpenAI response
            if 'choices' in response_data and len(response_data['choices']) > 0:
                analysis = response_data['choices'][0]['message']['content']
                
                self._set_headers(200)
                result = {
                    'analysis': analysis,
                    'success': True
                }
                self.wfile.write(json.dumps(result).encode())
            else:
                # Return error if there's an issue with the OpenAI response
                self._set_headers(200)
                response = {
                    'error': 'Failed to analyze image',
                    'analysis': 'OpenAI API provided an invalid response',
                    'success': False
                }
                self.wfile.write(json.dumps(response).encode())
                
        except Exception as e:
            print(f"Error analyzing image: {str(e)}")
            self._set_headers(500)
            response = {
                'error': f"Failed to analyze image: {str(e)}",
                'analysis': 'An error occurred during analysis',
                'success': False
            }
            self.wfile.write(json.dumps(response).encode())
        
    def do_GET(self):
        parsed_path = urllib.parse.urlparse(self.path)
        query_params = urllib.parse.parse_qs(parsed_path.query)
        
        if parsed_path.path == '/' or parsed_path.path == '/api':
            # Serve a basic HTML front-end for testing
            self._set_headers(200, 'text/html')
            html = """
            <!DOCTYPE html>
            <html lang="en">
            <head>
                <meta charset="UTF-8">
                <meta name="viewport" content="width=device-width, initial-scale=1.0">
                <title>Creately - Color Palette Generator</title>
                <style>
                    body {
                        font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif;
                        line-height: 1.6;
                        color: #333;
                        max-width: 800px;
                        margin: 0 auto;
                        padding: 20px;
                    }
                    h1 {
                        color: #2563EB;
                        margin-bottom: 10px;
                    }
                    h2 {
                        color: #4B5563;
                        margin-top: 30px;
                    }
                    .status {
                        background-color: #F3F4F6;
                        padding: 15px;
                        border-radius: 6px;
                        margin-bottom: 30px;
                    }
                    .palette-form {
                        background-color: #F9FAFB;
                        padding: 20px;
                        border-radius: 6px;
                        margin-bottom: 30px;
                    }
                    label {
                        display: block;
                        margin-bottom: 5px;
                        font-weight: 500;
                    }
                    input, select, textarea {
                        width: 100%;
                        padding: 8px;
                        margin-bottom: 15px;
                        border: 1px solid #D1D5DB;
                        border-radius: 4px;
                    }
                    button {
                        background-color: #2563EB;
                        color: white;
                        border: none;
                        padding: 10px 15px;
                        border-radius: 4px;
                        cursor: pointer;
                    }
                    button:hover {
                        background-color: #1D4ED8;
                    }
                    .palette-display {
                        display: flex;
                        margin-top: 20px;
                        height: 100px;
                    }
                    .color-box {
                        flex: 1;
                        margin-right: 2px;
                    }
                    .explanation {
                        margin-top: 15px;
                        padding: 10px;
                        background-color: #F3F4F6;
                        border-radius: 4px;
                    }
                    .color-code {
                        text-align: center;
                        margin-top: 5px;
                        font-family: monospace;
                    }
                </style>
            </head>
            <body>
                <h1>Creately - Color Palette Generator</h1>
                <p>An AI-powered creative platform that revolutionizes color palette generation through emotion-driven design solutions.</p>
                
                <div class="status">
                    <h3>API Status</h3>
                    <p>Database: <span id="db-status">Checking...</span></p>
                    <p>OpenAI API: <span id="openai-status">Checking...</span></p>
                </div>
                
                <div class="palette-form">
                    <h2>Generate Mood-Based Color Palette</h2>
                    <form id="palette-form">
                        <label for="mood">Mood:</label>
                        <select id="mood" name="mood" required>
                            <option value="happy">Happy</option>
                            <option value="calm">Calm</option>
                            <option value="energetic">Energetic</option>
                            <option value="professional">Professional</option>
                            <option value="romantic">Romantic</option>
                            <option value="mysterious">Mysterious</option>
                            <option value="playful">Playful</option>
                            <option value="elegant">Elegant</option>
                            <option value="bold">Bold</option>
                            <option value="serene">Serene</option>
                        </select>
                        
                        <label for="description">Additional Description (optional):</label>
                        <textarea id="description" name="description" rows="3" placeholder="Describe the feeling or context for more precise results..."></textarea>
                        
                        <button type="submit">Generate Palette</button>
                    </form>
                    
                    <div id="results" style="display: none;">
                        <h3>Your Palette</h3>
                        <div class="palette-display" id="palette-display"></div>
                        <div class="color-codes" id="color-codes"></div>
                        <div class="explanation" id="explanation"></div>
                    </div>
                </div>
                
                <script>
                    // Check API status on page load
                    fetch('/api')
                        .then(response => response.json())
                        .then(data => {
                            document.getElementById('db-status').textContent = data.database;
                            document.getElementById('openai-status').textContent = data.apiKeys.openai;
                        })
                        .catch(error => {
                            document.getElementById('db-status').textContent = 'Error checking status';
                            document.getElementById('openai-status').textContent = 'Error checking status';
                        });
                    
                    // Handle form submission
                    document.getElementById('palette-form').addEventListener('submit', function(e) {
                        e.preventDefault();
                        
                        const mood = document.getElementById('mood').value;
                        const description = document.getElementById('description').value;
                        
                        fetch('/api/palette/generate', {
                            method: 'POST',
                            headers: {
                                'Content-Type': 'application/json'
                            },
                            body: JSON.stringify({
                                mood: mood,
                                description: description,
                                count: 5
                            })
                        })
                        .then(response => response.json())
                        .then(data => {
                            displayPalette(data);
                        })
                        .catch(error => {
                            console.error('Error:', error);
                            alert('Failed to generate palette. Please try again.');
                        });
                    });
                    
                    function displayPalette(data) {
                        const paletteDisplay = document.getElementById('palette-display');
                        const colorCodes = document.getElementById('color-codes');
                        const explanation = document.getElementById('explanation');
                        
                        // Clear previous results
                        paletteDisplay.innerHTML = '';
                        colorCodes.innerHTML = '';
                        
                        // Display colors
                        data.colors.forEach(color => {
                            const colorBox = document.createElement('div');
                            colorBox.className = 'color-box';
                            colorBox.style.backgroundColor = color;
                            paletteDisplay.appendChild(colorBox);
                            
                            const colorCode = document.createElement('div');
                            colorCode.className = 'color-code';
                            colorCode.textContent = color;
                            colorCodes.appendChild(colorCode);
                        });
                        
                        // Display explanation
                        explanation.textContent = data.explanation;
                        
                        // Show results section
                        document.getElementById('results').style.display = 'block';
                    }
                </script>
            </body>
            </html>
            """
            self.wfile.write(html.encode())
        elif parsed_path.path == '/api/info':
            self._set_headers()
            response = {
                'message': 'Creately API (Python Fallback Server)',
                'timestamp': datetime.now().isoformat(),
                'status': 'SUCCESS',
                'database': 'Available (PostgreSQL)' if os.environ.get('DATABASE_URL') else 'Not configured',
                'apiKeys': {
                    'openai': 'Available' if os.environ.get('OPENAI_API_KEY') else 'Missing',
                    'pagespeed': 'Available' if os.environ.get('PAGESPEED_INSIGHTS_API_KEY') else 'Missing'
                },
                'server': 'Python Simple HTTP Server (Fallback)'
            }
            self.wfile.write(json.dumps(response, indent=2).encode())
        elif parsed_path.path == '/api/moods':
            self._set_headers()
            response = {
                'moods': [
                    'happy', 'calm', 'energetic', 'professional', 'romantic', 
                    'mysterious', 'playful', 'elegant', 'bold', 'serene',
                    'creative', 'modern', 'vintage', 'natural', 'futuristic',
                    'relaxed', 'vibrant', 'minimalist', 'luxurious', 'earthy'
                ]
            }
            self.wfile.write(json.dumps(response).encode())
        elif parsed_path.path.startswith('/api/pagespeed'):
            api_key = os.environ.get('PAGESPEED_INSIGHTS_API_KEY')
            self._set_headers(200 if api_key else 400)
            response = {
                'message': 'PageSpeed Insights API route',
                'timestamp': datetime.now().isoformat()
            }
            
            if not api_key:
                response['error'] = 'PageSpeed Insights API key is not configured'
            else:
                response['status'] = 'API key is configured'
                
            self.wfile.write(json.dumps(response, indent=2).encode())
        else:
            self._set_headers(404)
            response = {
                'error': 'Not Found',
                'message': f'The requested path {self.path} was not found',
                'timestamp': datetime.now().isoformat()
            }
            self.wfile.write(json.dumps(response, indent=2).encode())

# Read .env file and set environment variables
try:
    with open('.env', 'r') as env_file:
        for line in env_file:
            line = line.strip()
            if line and not line.startswith('#'):
                key, value = line.split('=', 1)
                os.environ[key] = value
    print("Environment variables loaded from .env file")
except Exception as e:
    print(f"Error loading .env file: {e}")

def run_server():
    Handler = SimpleHTTPRequestHandler
    httpd = socketserver.TCPServer(("0.0.0.0", PORT), Handler)
    print(f"Server running at http://0.0.0.0:{PORT}")
    httpd.serve_forever()

if __name__ == "__main__":
    run_server()

File: simple-server.js
/**
 * Simple Server Implementation
 * 
 * This is a lightweight implementation of the server for development
 * and testing purposes.
 */

const express = require('express');
const http = require('http');
const path = require('path');
const fs = require('fs');
const cors = require('cors');

// Configuration
const PORT = process.env.PORT || 3100;
const publicDir = path.join(__dirname, 'public');

// Create Express app
const app = express();
const server = http.createServer(app);

// Apply middleware
app.use(cors());
app.use(express.json());
app.use(express.urlencoded({ extended: true }));

// Enable logging
function logRequest(req, res, next) {
  const timestamp = new Date().toISOString();
  console.log(`[${timestamp}] ${req.method} ${req.url}`);
  next();
}

app.use(logRequest);

// Helper function to serve static files
function serveFile(filePath, res) {
  try {
    if (fs.existsSync(filePath)) {
      // Determine content type based on file extension
      const ext = path.extname(filePath).toLowerCase();
      const contentType = {
        '.html': 'text/html',
        '.js': 'text/javascript',
        '.css': 'text/css',
        '.json': 'application/json',
        '.png': 'image/png',
        '.jpg': 'image/jpeg',
        '.gif': 'image/gif',
        '.svg': 'image/svg+xml'
      }[ext] || 'text/plain';
      
      // Read and serve the file
      const data = fs.readFileSync(filePath);
      res.setHeader('Content-Type', contentType);
      res.end(data);
      return true;
    }
    return false;
  } catch (error) {
    console.error('Error serving file:', error.message);
    res.statusCode = 500;
    res.end('Internal server error');
    return true;
  }
}

// Function to handle API requests
function handleApiRequest(req, res) {
  const apiPath = req.url.substring('/api/'.length);
  
  // Simple API endpoints
  if (apiPath === 'status') {
    res.setHeader('Content-Type', 'application/json');
    res.end(JSON.stringify({
      status: 'ok',
      timestamp: new Date().toISOString(),
      version: '1.0.0'
    }));
    return true;
  }
  
  // Mock AI service endpoints
  if (apiPath.startsWith('ai/')) {
    const aiEndpoint = apiPath.substring('ai/'.length);
    
    if (aiEndpoint === 'status') {
      res.setHeader('Content-Type', 'application/json');
      res.end(JSON.stringify({
        available: true,
        providers: ['mock-provider'],
        default: 'mock-provider'
      }));
      return true;
    }
    
    // Handle POST requests to AI endpoints
    if (req.method === 'POST') {
      let body = '';
      
      req.on('data', chunk => {
        body += chunk.toString();
      });
      
      req.on('end', () => {
        let data;
        try {
          data = JSON.parse(body);
        } catch (error) {
          res.statusCode = 400;
          res.setHeader('Content-Type', 'application/json');
          res.end(JSON.stringify({ error: 'Invalid JSON' }));
          return;
        }
        
        // Generate mock responses for different AI endpoints
        let response;
        
        if (aiEndpoint === 'generate-text') {
          response = {
            content: `Mock response for: ${data.prompt || 'No prompt provided'}`,
            provider: 'mock-provider',
            model: 'mock-model',
            usage: {
              promptTokens: 10,
              completionTokens: 20,
              totalTokens: 30
            }
          };
        } else if (aiEndpoint === 'generate-json') {
          response = {
            json: {
              result: "Mock JSON response",
              items: [1, 2, 3],
              success: true
            },
            provider: 'mock-provider',
            model: 'mock-model'
          };
        } else if (aiEndpoint === 'chat') {
          response = {
            message: {
              role: 'assistant',
              content: `Mock chat response to: ${data.messages ? data.messages.map(m => m.content).join(', ') : 'No messages provided'}`
            },
            provider: 'mock-provider',
            model: 'mock-model'
          };
        } else {
          response = {
            error: 'Unknown AI endpoint'
          };
        }
        
        res.statusCode = 200;
        res.setHeader('Content-Type', 'application/json');
        res.end(JSON.stringify(response));
      });
      
      return true;
    }
  }
  
  return false;
}

// Create public directory if it doesn't exist
if (!fs.existsSync(publicDir)) {
  fs.mkdirSync(publicDir, { recursive: true });
  
  // Create a basic index.html if it doesn't exist
  const indexPath = path.join(publicDir, 'index.html');
  if (!fs.existsSync(indexPath)) {
    const htmlContent = `
      <!DOCTYPE html>
      <html>
        <head>
          <title>Simple Server</title>
          <style>
            body {
              font-family: Arial, sans-serif;
              max-width: 800px;
              margin: 0 auto;
              padding: 20px;
              line-height: 1.6;
            }
            h1 { color: #2c3e50; }
            .status { color: #27ae60; font-weight: bold; }
            .endpoints { margin-top: 20px; }
            code {
              background: #f8f8f8;
              padding: 2px 4px;
              border-radius: 3px;
              font-family: monospace;
            }
          </style>
        </head>
        <body>
          <h1>Simple Server Running</h1>
          <p>Status: <span class="status">Online</span></p>
          
          <div class="endpoints">
            <h2>Available Endpoints:</h2>
            <ul>
              <li><code>GET /api/status</code> - Server status</li>
              <li><code>GET /api/ai/status</code> - AI service status</li>
              <li><code>POST /api/ai/generate-text</code> - Generate text</li>
              <li><code>POST /api/ai/generate-json</code> - Generate JSON</li>
              <li><code>POST /api/ai/chat</code> - Chat completion</li>
            </ul>
          </div>
          
          <div id="testArea">
            <h2>Test Area</h2>
            <button id="testStatusBtn">Test Status</button>
            <button id="testTextBtn">Test Text Generation</button>
            <button id="testChatBtn">Test Chat Completion</button>
            <pre id="result" style="background: #f8f8f8; padding: 10px; margin-top: 10px; border-radius: 5px;"></pre>
          </div>
          
          <script>
            // Simple test functions
            async function testStatus() {
              try {
                const response = await fetch('/api/status');
                const data = await response.json();
                document.getElementById('result').textContent = JSON.stringify(data, null, 2);
              } catch (error) {
                document.getElementById('result').textContent = 'Error: ' + error.message;
              }
            }
            
            async function testTextGeneration() {
              try {
                const response = await fetch('/api/ai/generate-text', {
                  method: 'POST',
                  headers: {
                    'Content-Type': 'application/json'
                  },
                  body: JSON.stringify({
                    prompt: 'Hello, world!'
                  })
                });
                const data = await response.json();
                document.getElementById('result').textContent = JSON.stringify(data, null, 2);
              } catch (error) {
                document.getElementById('result').textContent = 'Error: ' + error.message;
              }
            }
            
            async function testChatCompletion() {
              try {
                const response = await fetch('/api/ai/chat', {
                  method: 'POST',
                  headers: {
                    'Content-Type': 'application/json'
                  },
                  body: JSON.stringify({
                    messages: [
                      { role: 'system', content: 'You are a helpful assistant.' },
                      { role: 'user', content: 'Hello, how are you?' }
                    ]
                  })
                });
                const data = await response.json();
                document.getElementById('result').textContent = JSON.stringify(data, null, 2);
              } catch (error) {
                document.getElementById('result').textContent = 'Error: ' + error.message;
              }
            }
            
            // Add event listeners
            document.getElementById('testStatusBtn').addEventListener('click', testStatus);
            document.getElementById('testTextBtn').addEventListener('click', testTextGeneration);
            document.getElementById('testChatBtn').addEventListener('click', testChatCompletion);
          </script>
        </body>
      </html>
    `;
    fs.writeFileSync(indexPath, htmlContent);
    console.log(`Created ${indexPath}`);
  }
}

// Handle all requests
app.use((req, res) => {
  // Handle API requests
  if (req.url.startsWith('/api/')) {
    if (handleApiRequest(req, res)) {
      return;
    }
    
    // If API endpoint not found
    res.statusCode = 404;
    res.setHeader('Content-Type', 'application/json');
    res.end(JSON.stringify({ error: 'API endpoint not found' }));
    return;
  }
  
  // Handle static file requests
  let filePath = path.join(publicDir, req.url === '/' ? 'index.html' : req.url);
  
  // Remove query parameters
  filePath = filePath.split('?')[0];
  
  if (serveFile(filePath, res)) {
    return;
  }
  
  // Try to serve index.html for any other routes (SPA support)
  const indexPath = path.join(publicDir, 'index.html');
  if (serveFile(indexPath, res)) {
    return;
  }
  
  // If all else fails, return 404
  res.statusCode = 404;
  res.end('Not found');
});

// Start the server
server.listen(PORT, '0.0.0.0', () => {
  console.log(`Simple server running on http://localhost:${PORT}`);
});

// Handle graceful shutdown
process.on('SIGTERM', () => {
  console.log('SIGTERM received, shutting down gracefully');
  server.close(() => {
    console.log('HTTP server closed');
    process.exit(0);
  });
});

process.on('SIGINT', () => {
  console.log('SIGINT received, shutting down gracefully');
  server.close(() => {
    console.log('HTTP server closed');
    process.exit(0);
  });
});

module.exports = { app, server };